processor/src/test/java/com/linecorp/decaton/processor/runtime/ProcessorSubscriptionTest.java

```diff
@@ -170,4 +183,61 @@ public void testOffsetRegression() throws Exception {
         OffsetAndMetadata offset = offsets.get(tp);
         assertEquals(100L, offset.offset());
     }
+
+    @Test(timeout = 10000L)
+    public void testTerminateAsync() throws Exception {
+        TopicPartition tp = new TopicPartition("topic", 0);
+        DecatonMockConsumer consumer = new DecatonMockConsumer() {
+            @Override
+            public synchronized ConsumerRecords<String, byte[]> poll(Duration timeout) {
+                rebalanceListener.onPartitionsAssigned(assignment());
+                return super.poll(timeout);
+            }
+        };
+        consumer.updateEndOffsets(singletonMap(tp, 10L));
+        ExecutorService executor = Executors.newSingleThreadExecutor();
+        CountDownLatch subscribed = new CountDownLatch(1);
+        CountDownLatch asyncProcessingStarted = new CountDownLatch(1);
+        CountDownLatch letTasksComplete = new CountDownLatch(1);
+        DecatonProcessor<String> processor = (context, task) -> {
+            final DeferredCompletion completion = context.deferCompletion();
+            executor.submit(() -> {
+                try {
+                    letTasksComplete.await();
+                } catch (InterruptedException e) {
+                    throw new RuntimeException(e);
+                }
+                completion.complete();
+            });
+            asyncProcessingStarted.countDown();
+        };
+        SubscriptionScope scope = scope(tp.topic(), 9000L);
+        final ProcessorSubscription subscription = new ProcessorSubscription(
+                scope,
+                () -> consumer,
+                ProcessorsBuilder.consuming(scope.topic(),
+                                            (byte[] bytes) -> new DecatonTask<>(
+                                                    TaskMetadata.builder().build(), "dummy", bytes))
+                                 .thenProcess(processor)
+                                 .build(null),
+                scope.props(),
+                newState -> {
+                    if (newState == State.RUNNING) {
+                        subscribed.countDown();
+                    }
+                });
+        subscription.start();
+        subscribed.await();
+        consumer.rebalance(singleton(tp));
+        consumer.addRecord(new ConsumerRecord<>(tp.topic(), tp.partition(), 10, "", NO_DATA));
+        consumer.addRecord(new ConsumerRecord<>(tp.topic(), tp.partition(), 11, "", NO_DATA));
+        asyncProcessingStarted.await();
+        subscription.initiateShutdown();
+        assertTrue(consumer.committed(singleton(tp)).isEmpty());
+        letTasksComplete.countDown();
```

Comment:
hmm, is this working as a case for this https://github.com/line/decaton/pull/79#discussion_r572580818 ?

I think in this implementation, 2 tasks are first fed into `DecatonProcessor` anyway (because they returns immediately) and decaton's internal queue quickly becomes empty, so not testing that "decaton processes tasks that are fetched but still in queue at the beginning shutdown" I think?

Reply:
The first task can't complete until `letTasksComplete.countDown();` is called, so the second task will still be queued. I'll add an assert that checks they're still pending at this point?

Reply:
Decaton gives the 2nd task to `DecatonProcessor#process` even when the 1st task hasn't yet completed thought?
When a processor uses `DeferredCompletion`, it is processor's responsibility to serialize execution of multiple tasks with the same key for after `#process` returns.

Reply:
Ah indeed. I've added blocking in the synchronous part of processing as well so that we test that case.